radicle-tracker/src/thread.rs

```diff
@@ -0,0 +1,827 @@
+        let replies_count = self.replies_count();
+        let main_bound = self.main_thread.len() - 1;
+
+        let replies_bound = if replies_count == 0 {
+            None
+        } else {
+            Some(replies_count - 1)
+        };
+
+        match self._finger {
+            Finger::Root => {
+                if self.main_thread.is_empty() {
+                    return Err(Error::NextMainOutOfBounds);
+                }
+
+                self._finger = Finger::Main(0);
+                Ok(())
+            },
+            Finger::Main(ref mut main_ix) => match reply_to {
+                ReplyTo::Main => {
+                    if *main_ix == main_bound {
+                        return Err(Error::NextMainOutOfBounds);
+                    }
+
+                    *main_ix += 1;
+                    Ok(())
+                },
+                ReplyTo::Thread => match replies_bound {
+                    None => Err(Error::NextRepliesOutOfBound),
+                    // We're ensuring that we have replies
+                    Some(_) => {
+                        // We start at one because the replies are the tail
+                        // of the non-empty vec in Replies
+                        self._finger = Finger::Thread((*main_ix, 1));
+                        Ok(())
+                    },
+                },
+            },
+            Finger::Thread((ref mut main_ix, ref mut replies_ix)) => match reply_to {
+                ReplyTo::Main => {
+                    if *main_ix == main_bound {
+                        return Err(Error::NextMainOutOfBounds);
+                    }
+
+                    self._finger = Finger::Main(*main_ix + 1);
+                    Ok(())
+                },
+                ReplyTo::Thread => match replies_bound {
+                    None => Err(Error::NextRepliesOutOfBound),
+                    Some(bound) => {
+                        if *replies_ix == bound {
+                            return Err(Error::NextRepliesOutOfBound);
+                        } else {
+                            *replies_ix += 1;
+                        }
+                        Ok(())
+                    },
+                },
+            },
+        }
+    }
+
+    /// Look at the root of the thread.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::error::Error;
+    /// #
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// use radicle_tracker::{ReplyTo, Status, Thread};
+    ///
+    /// let (mut thread) = Thread::new(String::from("Discussing rose trees"));
+    ///
+    /// // Reply to the main thread
+    /// thread.reply(String::from("I love rose trees!"), ReplyTo::Main);
+    ///
+    /// thread.root();
+    /// assert_eq!(thread.view(), Ok(&Status::Live(String::from("Discussing rose trees"))));
+    /// #
+    /// #     Ok(())
+    /// # }
+    /// ```
+    pub fn root(&mut self) {
+        self._finger = ROOT_FINGER;
+    }
```

Comment:
Based on the name, I kind of expect this to return the root, since `root` is one of the fields, and a thread has a root. One suggestion would be to have instead a function like `Thread::navigate_to(finger)` (or call it `goto`), and pass in `Finger::Root` to goto the root.

Reply:
This _does_ return to the root. I was thinking about adding a `navigate_to` (HAH! `goto`), but it'll be a similar concept of assigning the `self._finger`. Having a `root` function is still useful.

Reply:
"return the root" != "return *to* the root" :grin: 

What I find confusing is that the thread has a root and a field called root, yet the method called root isn't a getter, it's actually something that does internal mutation.

Reply:
Fair, the language level semantics could probably be improved :sweat_smile: 

> What I find confusing is that the Thread has a root and a field called root, yet the method called root isn't a getter for that field (as you would expect), it's actually something that does internal mutation.

That's true... I guess the intention is that `thread.root().view() === thread.root`, but also we want the user to be able to navigate back to the root to execute further navigation from there.

Again, it's probably just the language of the code that needs work :thinking: